src/daemon/keymap.c

```diff
@@ -609,6 +609,7 @@ void process_input_urb(void* context, unsigned char* buffer, int urblen, ushort
                         } else {
                             kb->input.rel_x += (buffer[5] << 8) | buffer[4];
                             kb->input.rel_y += (buffer[7] << 8) | buffer[6];
+                            kb->input.whl_rel_y = (signed char)buffer[8];
```

Comment:
I think this might cause issues with other bragi devices. I need to look into it.

Reply:
So, having finally looked at how my mouse behaves, in software mode, for every scroll event, it spits out two packets.
```
[I] ckb1 Input Recv: 00 02 00 00 ff 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
[I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff 
```
The bottom one is the one that this change receives the scroll information from, which is a hardware mode packet.
The scroll information should be read from the software mode packet, which is the first one in this case.
Does your mouse not emit that packet? Does it differ in some way from mine?

Reply:
My katar pro xt with my patchset on top of the master branch from where it was based on does not generate the 1st packet for the scroll wheel, it does generate these for the buttons though. I had ckb-next running so i presume it was in software mode

<pre>
Sep 29 21:27:47 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 29 21:27:47 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff
Sep 29 21:27:48 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 29 21:27:48 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff
Sep 29 21:27:48 : [I] ckb1 Input Recv: 00 00 00 00 01 00 00 00 00
Sep 29 21:27:49 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 29 21:27:49 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff
Sep 29 21:27:50 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 29 21:27:50 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff
Sep 29 21:27:51 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 29 21:27:51 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff
Sep 29 21:27:51 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 29 21:27:52 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff
Sep 29 21:27:56 : [I] ckb1 Input Recv: 00 02 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 29 21:27:56 : [I] ckb1 Input Recv: 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 29 21:27:57 : [I] ckb1 Input Recv: 00 02 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 29 21:27:57 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 00
Sep 29 21:27:57 : [I] ckb1 Input Recv: 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 29 21:27:57 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 00
Sep 29 21:27:57 : [I] ckb1 Input Recv: 00 00 00 00 01 00 00 00 00
Sep 29 21:27:58 : [I] ckb1 Input Recv: 00 02 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 29 21:27:58 : [I] ckb1 Input Recv: 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 29 21:27:59 : [I] ckb1 Input Recv: 00 02 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 29 21:27:59 : [I] ckb1 Input Recv: 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 29 21:28:01 : [I] ckb1 Input Recv: 00 02 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 29 21:28:01 : [I] ckb1 Input Recv: 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 29 21:28:03 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff
Sep 29 21:28:04 : [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 29 21:28:08 : [I] ckb1 Input Recv: 00 00 00 00 01 00 00 00 00
</pre>
in fact i tried `zgrep 'ckb1 Input Recv' /var/log/syslog* | grep -v ': .. .. .. .. .. .. .. .. ..$' | grep -v repeated | grep ff` and this has 0 matches.
Of course i know very little about the codebase and the protocol so its quite possible iam missing something

Reply:
This seems frustrating. Do you also not get both packets when you press a button?
For example, when I left click in software mode, I get:
```
[I] ckb1 Input Recv: 00 02 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
[I] ckb1 Input Recv: 01 00 00 00 00 00 00 00 00 
```
but in hardware mode I get only:
```
[I] ckb1 Input Recv: 01 00 00 00 00 00 00 00 00 
```

> I had ckb-next running so i presume it was in software mode

As long as the GUI is running and controlling the lights, it's in software mode. Quitting the GUI using either the tray icon or the button under settings should switch the device back to hardware mode.

Reply:
It seems it doesnt follow the expected behavior.
I also rebased the code locally on top of current master to make sure iam not missing some bugfix

<pre>
wheel and button test
Sep 30 22:51:16  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 01 00 00 00 00
Sep 30 22:51:17  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 01 00 00
Sep 30 22:51:20  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 30 22:51:20  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 01 00 00
Sep 30 22:51:21  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff
Sep 30 22:51:22  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 30 22:51:22  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 01 00 00
Sep 30 22:51:22  ckb-next-daemon[2842705]: message repeated 3 times: [ [I] ckb1 Input Recv: 00 00 00 00 00 00 01 00 00]
Sep 30 22:51:22  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 ff ff 00 00 00
Sep 30 22:51:22  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 01 00 00
Sep 30 22:51:23  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff
Sep 30 22:51:23  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 00
Sep 30 22:51:23  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 01 00 00
Sep 30 22:51:23  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 30 22:51:24  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 01 00 00
Sep 30 22:51:24  ckb-next-daemon[2842705]: message repeated 2 times: [ [I] ckb1 Input Recv: 00 00 00 00 00 00 01 00 00]
Sep 30 22:51:24  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff
Sep 30 22:51:25  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 30 22:51:25  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 01 00 00
Sep 30 22:51:25  ckb-next-daemon[2842705]: message repeated 8 times: [ [I] ckb1 Input Recv: 00 00 00 00 00 00 01 00 00]
Sep 30 22:51:25  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 ff ff 00 00 00
Sep 30 22:51:25  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 01 00 00
Sep 30 22:51:25  ckb-next-daemon[2842705]: message repeated 4 times: [ [I] ckb1 Input Recv: 00 00 00 00 00 00 01 00 00]
Sep 30 22:51:25  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff
Sep 30 22:51:26  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 01 00 00 00 00 00 00 00 00
Sep 30 22:51:27  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 00
Sep 30 22:51:27  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 01 00 00 00 00 00 00 00 00
Sep 30 22:51:27  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 00
Sep 30 22:51:28  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 02 00 00 00 00 00 00 00 00

ckb-next is started, GUI is vissible from here on
wheel and button test is repeated

Sep 30 22:52:18  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 ff ff 00
Sep 30 22:52:48  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 30 22:52:49  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff
Sep 30 22:52:49  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 30 22:52:50  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff
Sep 30 22:52:51  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 30 22:52:51  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 ff
Sep 30 22:52:52  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 00 00 00 00 00 00 00 01
Sep 30 22:52:52  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 02 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 30 22:52:53  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 30 22:52:53  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 02 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 30 22:52:53  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 30 22:52:55  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 02 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 30 22:52:55  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 30 22:52:56  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 02 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 30 22:52:56  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 30 22:53:04  ckb-next-daemon[2842705]: [I] ckb1 Input Recv: 00 12 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
</pre>

Reply:
Please try running it outside of systemd.
You should be able to `sudo systemctl stop ckb-next-daemon` and manually `sudo ckb-next-daemon`.

I don't think this is the case, but it might be rate limiting the logs somehow and maybe we're missing some of them, but it seems unlikely. Just want to be sure.

Reply:
Tried without systemctl, same behavior

Reply:
Understood. I guess for now we can ignore the scroll data on the other packet and see how it goes.